backend/substrapp/tasks/tasks.py

```diff
@@ -323,7 +326,9 @@ def prepare_materials(subtuple, tuple_type):
     put_metric(subtuple_directory, metrics_content)
     put_algo(subtuple_directory, algo_content)
     if tuple_type == 'testtuple':
-        put_model(subtuple, subtuple_directory, model_content)
+        traintuple_metadata = get_traintuple_metadata(subtuple['traintupleKey'])
+        model_hash = traintuple_metadata['outModel']['hash']
+        put_model(subtuple, subtuple_directory, model_content, model_hash)
```

Comment:
Be careful, you change the behavior of the previous flow.
Here you're making a call to the ledger to get a info that it was (before https://github.com/SubstraFoundation/substra-chaincode/pull/21) in the testtuple structure.
Is it something that we want ?

Reply:
@Kelvin-M That is intentional (and actually the whole point of this PR!). The rationale for changing the current behavior is that since composite testtuples have 2 models (and not just one), the current testtuple structure (only one model) doesn't work in the composite case.

The approach I propose solves the problem by making testtuples ignorant of the type of model(s) they work with (regular or composite). This might not be the best approach: I'm happy to hear suggestions :)

The main benefit of the proposed solution is that it prevents the introduction of a new `compositeTesttuple` type in the chaincode, which would be essentially a big copy paste of `testTuple` (with a few slight modifications). That would translate to hundreds if not thousands of new lines of code, we'd have to remember to keep the two structures in sync, test them both, etc. In the long term, that's a significant cost.

Regarding the extra ledger call, I think this could be mitigated in future work with caching on the backend (cache the model hash and storage address associated with a traintuple key). AFAIK, the model hash/address never change for a given traintuple.

cc @inalgnu @thibaultrobert since you're most involved in this

Reply:
Okay :) 

> The approach I propose solves the problem by making testtuples ignorant of the type of model(s) they work with (regular or composite). This might not be the best approach: I'm happy to hear suggestions :)

Okay, I think the generic tuple PR will need extra work to make it compatible with it as `put_model` work only with one model ?

I think it will be a refacto in generic tuple because of copy/paste code that already happen as we can see with `CompositeAlgo` :)

 Is there a reason for this fix to not be directly included in generic tuple PR, it seems to be very link to it ?

Reply:
@Kelvin-M I think the required chaincode bits have already been implemented as part of the [chaincode PR](https://github.com/SubstraFoundation/substra-chaincode/pull/5): in particular `getCompositeTraintuple` in addition to `getTraintuple`.

The current plan is to have a `traintupleType` field in the `outputTesttuple` returned by the ledger. Depending on this field, the backend can request the ledger for either a regular model or a composite model.

I think the plan is to split `put_model` into a composite and non composite versions (see [this part of the backend composite PR](https://github.com/SubstraFoundation/substra-backend/pull/13/files#diff-1c7f95fdd9883397a5489522f326ccb7R222).

> Is there a reason for this fix to not be directly included in generic tuple PR, it seems to be very link to it ?

My rationale is that since this PR is preparatory work, and it can be be merged into `dev` independently, doing so would reduce the complexity of the (already huge) composite PR. Merging it early would allow us to see any issues with it earlier rather than later.

Reply:
Okay, I will try to test it in multiple setup to see if it's good :)